source/extensions/quic/connection_id_generator/quic_lb/quic_lb.cc

```diff
@@ -0,0 +1,353 @@
+                Server::Configuration::FactoryContext& context) {
+  std::unique_ptr<Factory> ret(new Factory(config));
+
+  auto server_id_or_result =
+      Config::DataSource::read(config.server_id(), false, context.serverFactoryContext().api());
+  RETURN_IF_NOT_OK_REF(server_id_or_result.status());
+
+  std::string server_id = server_id_or_result.value();
+
+  // Create a test instance using all the same parameters, but with a fake key (because we don't
+  // have the real key yet) to surface any errors while we're still in the config loading stage.
+  {
+    auto test_instance_or_result =
+        QuicLbConnectionIdGenerator::ThreadLocalData::create(config, server_id);
+    RETURN_IF_NOT_OK_REF(test_instance_or_result.status());
+    std::string test_key(quic::kLoadBalancerKeyLen, '0');
+    auto result = test_instance_or_result.value()->updateKeyAndVersion(test_key, 0);
+    RETURN_IF_NOT_OK_REF(result);
+
+    // Record the length of the encoder's connection ID so that we know the offset to look for
+    // routing information.
+    ret->encoder_connection_id_length_ =
+        test_instance_or_result.value()->encoder_.GenerateConnectionId().length();
+  }
+
+  ret->secrets_provider_ =
+      secretsProvider(config.encryption_parmeters(), context.getTransportSocketFactoryContext(),
+                      context.initManager());
+  if (ret->secrets_provider_ == nullptr) {
+    return absl::InvalidArgumentError("invalid encryption_parmeters config");
+  }
+
+  ret->secrets_provider_validation_callback_handle_ = ret->secrets_provider_->addValidationCallback(
+      [&api = context.serverFactoryContext().api()](
+          const envoy::extensions::transport_sockets::tls::v3::GenericSecret& secret)
+          -> absl::Status { return getAndValidateKeyAndVersion(secret, api).status(); });
+
+  ret->secrets_provider_update_callback_handle_ = ret->secrets_provider_->addUpdateCallback(
+      [&factory = *ret, &api = context.serverFactoryContext().api()]() -> absl::Status {
+        const envoy::extensions::transport_sockets::tls::v3::GenericSecret* secret =
+            factory.secrets_provider_->secret();
+        if (secret == nullptr) {
+          return absl::NotFoundError("secret update callback called with empty secret");
+        }
+
+        auto data_or_result = getAndValidateKeyAndVersion(*secret, api);
+        RETURN_IF_NOT_OK_REF(data_or_result.status());
+
+        factory.tls_slot_->runOnAllThreads(
+            [data =
+                 data_or_result.value()](OptRef<QuicLbConnectionIdGenerator::ThreadLocalData> obj) {
+              ASSERT(obj.has_value(),
+                     "Guaranteed if `set()` was previously called on the tls slot");
+
+              // TODO: error handling from both of these calls
+              auto result = obj->updateKeyAndVersion(data.first, data.second);
+              ASSERT(result.ok(), "Parameters and data was already validated");
```

Comment:
When update config fails, can we keep using the old config and not to crash?

Reply:
I changed this to ENVOY_BUG. This should be unreachable, but now if it does fail it'll log and continue using the old config.